### PR TITLE
[WIP][SPARK-27856][SQL] Only allow type upcasting when inserting table

### PR DESCRIPTION
**Pull request by gengliangwang**
## What changes were proposed in this pull request?

In Spark version 2.4 and earlier, when inserting into a table, Spark will cast the data type of input query to the data type of target table by coercion. This can be super confusing if users make a mistake and write string values to an int column. Users won't get error/warning and see null values in the target table.

Since Spark 3.0, by default only upcasting is allowed when inserting data into table. E.g. `int` -> `long` and `int` -> `string` are allowed, while `long` -> `int` or `string` -> `int` are not allowed. The old behaviour is preserved under a newly added configuration `spark.sql.legacy.insertUnsafeCasts` with a default value of `false`.

## How was this patch tested?

Unit test

### CODE REVIEW

**Comment by gengliangwang** [comment]
As @cloud-fan is taking a vacation for weeks, I am taking https://github.com/apache/spark/pull/24721 over. I have updated the SQL configuration naming, and updated migration guide. 

@gatorsmile @ueshin @rdblue  Please help review it.

File: docs/sql-migration-guide-upgrade.md [hunk 1]
```diff
@@ -50,6 +50,8 @@ license: |
 
   - In Spark version 2.4 and earlier, JSON datasource and JSON functions like `from_json` convert a bad JSON record to a row with all `null`s in the PERMISSIVE mode when specified schema is `StructType`. Since Spark 3.0, the returned row can contain non-`null` fields if some of JSON column values were parsed and converted to desired types successfully.
 
+  - In Spark version 2.4 and earlier, when inserting into a table, Spark will cast the data type of input query to the data type of target table by coercion. Since Spark 3.0, by default only upcasting is allowed when inserting data into table. E.g. `int` -> `long` and `int` -> `string` are allowed, while `long` -> `int` or `string` -> `int` are not allowed. The old behaviour is preserved under a newly added configuration `spark.sql.legacy.insertTable.typeCoercion` with a default value of `false`.
```
Comment (rdblue) [inline docs/sql-migration-guide-upgrade.md]: I think the configuration name should be more clear. The name `typeCoercion` doesn't describe what's happening. How about `spark.sql.legacy.insertUnsafeCasts`? I think that is clear about what will actually happen when it is enabled.

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/analysis/ResolveOutputRelation.scala [hunk 2]
```diff
@@ -0,0 +1,146 @@
...
+
+/**
+ * Resolves columns of an output table from the data in a logical plan. This rule will:
+ *
+ * - Reorder columns when the write is by name
+ * - Insert safe casts when data types do not match
+ * - Insert aliases when column names do not match
+ * - Detect plans that are not compatible with the output table and throw AnalysisException
+ */
+object ResolveOutputRelation extends Rule[LogicalPlan] {
```
Comment (rdblue) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/analysis/ResolveOutputRelation.scala]: Why was this moved?

It is difficult to see whether anything changed in this class. If the move was not required, please move it back.

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/Cast.scala [hunk 3]
```diff
@@ -126,10 +126,14 @@ object Cast {
...
     case _ if from == to => true
+    case (NullType, _) => true
+
     case (from: NumericType, to: DecimalType) if to.isWiderThan(from) => true
     case (from: DecimalType, to: NumericType) if from.isTighterThan(to) => true
-    case (f, t) if legalNumericPrecedence(f, t) => true
+    case (f: NumericType, t: NumericType) if legalNumericPrecedence(f, t) => true
+
     case (DateType, TimestampType) => true
+
```
Comment (rdblue) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/Cast.scala]: Nit: please remove non-functional whitespace additions.

File: sql/catalyst/src/test/scala/org/apache/spark/sql/catalyst/encoders/EncoderResolutionSuite.scala [hunk 4]
```diff
@@ -230,10 +230,11 @@ class EncoderResolutionSuite extends PlanTest {
   castSuccess[Long, String]
   castSuccess[Int, java.math.BigDecimal]
   castSuccess[Long, java.math.BigDecimal]
+  castSuccess[java.math.BigDecimal, Double]
```
Comment (rdblue) [inline sql/catalyst/src/test/scala/org/apache/spark/sql/catalyst/encoders/EncoderResolutionSuite.scala]: Why is this cast now allowed? This is not a safe cast because the scale of the decimal is lost.

File: sql/core/src/main/scala/org/apache/spark/sql/execution/datasources/rules.scala [hunk 5]
```diff
@@ -356,8 +358,28 @@ case class PreprocessTableInsertion(conf: SQLConf) extends Rule[LogicalPlan] {
           s"including ${staticPartCols.size} partition column(s) having constant value(s).")
     }
 
-    val newQuery = DDLPreprocessingUtils.castAndRenameQueryOutput(
-      insert.query, expectedColumns, conf)
+    val newQuery = if (conf.getConf(SQLConf.LEGACY_INSERT_TABLE_TYPE_COERCION)) {
+      DDLPreprocessingUtils.castAndRenameQueryOutput(insert.query, expectedColumns, conf)
+    } else {
+      val errors = new mutable.ArrayBuffer[String]()
```
Comment (rdblue) [inline sql/core/src/main/scala/org/apache/spark/sql/execution/datasources/rules.scala]: Why not run `ResolveOutputRelation`?

File: sql/hive/compatibility/src/test/scala/org/apache/spark/sql/hive/execution/HiveCompatibilitySuite.scala [hunk 6]
```diff
@@ -59,6 +59,8 @@ class HiveCompatibilitySuite extends HiveQueryFileTest with BeforeAndAfter {
     TestHive.setConf(SQLConf.IN_MEMORY_PARTITION_PRUNING, true)
     // Ensures that cross joins are enabled so that we can test them
     TestHive.setConf(SQLConf.CROSS_JOINS_ENABLED, true)
+    // Force type case during table insertion
+    TestHive.setConf(SQLConf.LEGACY_INSERT_TABLE_TYPE_COERCION, true)
```
Comment (rdblue) [inline sql/hive/compatibility/src/test/scala/org/apache/spark/sql/hive/execution/HiveCompatibilitySuite.scala]: Why is this required?

File: sql/hive/src/test/scala/org/apache/spark/sql/hive/execution/HiveQuerySuite.scala [hunk 7]
```diff
@@ -374,15 +374,15 @@ class HiveQuerySuite extends HiveComparisonTest with SQLTestUtils with BeforeAnd
       sql(s"CREATE TABLE test_partition (a STRING) PARTITIONED BY (b BIGINT, c STRING)")
       sql(s"CREATE TABLE ptest (a STRING, b BIGINT, c STRING)")
 
-      val analyzedPlan = sql(
+      val optimizedPlan = sql(
```
Comment (rdblue) [inline sql/hive/src/test/scala/org/apache/spark/sql/hive/execution/HiveQuerySuite.scala]: Why does this use the optimized plan now? So that `checkAnalysis` will run?

File: sql/hive/src/test/scala/org/apache/spark/sql/hive/execution/HiveSerDeReadWriteSuite.scala [hunk 8]
```diff
@@ -52,9 +52,9 @@ class HiveSerDeReadWriteSuite extends QueryTest with SQLTestUtils with TestHiveS
   private def checkNumericTypes(fileFormat: String, dataType: String, value: Any): Unit = {
     withTable("hive_serde") {
       hiveClient.runSqlHive(s"CREATE TABLE hive_serde (c1 $dataType) STORED AS $fileFormat")
-      hiveClient.runSqlHive("INSERT INTO TABLE hive_serde values(1)")
+      hiveClient.runSqlHive(s"INSERT INTO TABLE hive_serde VALUES(1)")
```
Comment (rdblue) [inline sql/hive/src/test/scala/org/apache/spark/sql/hive/execution/HiveSerDeReadWriteSuite.scala]: Please remove all non-functional changes like altering the case of keywords.

File: sql/hive/src/test/scala/org/apache/spark/sql/hive/execution/HiveSerDeReadWriteSuite.scala [hunk 9]
```diff
@@ -52,9 +52,9 @@ class HiveSerDeReadWriteSuite extends QueryTest with SQLTestUtils with TestHiveS
...
       checkAnswer(spark.table("hive_serde"), Row(1))
-      spark.sql(s"INSERT INTO TABLE hive_serde values($value)")
+      spark.sql(s"INSERT INTO TABLE hive_serde VALUES(CAST($value AS $dataType))")
```
Comment (rdblue) [inline sql/hive/src/test/scala/org/apache/spark/sql/hive/execution/HiveSerDeReadWriteSuite.scala]: It's fine to fix this one since the line is already being modified.

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/analysis/ResolveOutputRelation.scala (hunk 2 above)
Comment (gengliangwang) [reply sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/analysis/ResolveOutputRelation.scala]: > move it outside of the Analyzer class, so that we can call its methods.

As per @cloud-fan commented in https://github.com/apache/spark/pull/24721/files#r287800626

File: sql/core/src/main/scala/org/apache/spark/sql/execution/datasources/rules.scala (hunk 5 above)
Comment (gengliangwang) [reply sql/core/src/main/scala/org/apache/spark/sql/execution/datasources/rules.scala]: Here is for the V1 path. `InsertIntoTable` won't be matched in `ResolveOutputRelation`.

File: sql/hive/src/test/scala/org/apache/spark/sql/hive/execution/HiveQuerySuite.scala (hunk 7 above)
Comment (gengliangwang) [reply sql/hive/src/test/scala/org/apache/spark/sql/hive/execution/HiveQuerySuite.scala]: The `Cast` operator will be eliminated by optimization rule `ConstantFolding`.

File: sql/hive/compatibility/src/test/scala/org/apache/spark/sql/hive/execution/HiveCompatibilitySuite.scala (hunk 6 above)
Comment (gengliangwang) [reply sql/hive/compatibility/src/test/scala/org/apache/spark/sql/hive/execution/HiveCompatibilitySuite.scala]: Not sure about that. Actually, I don't know how to run the test suite

**Comment by gengliangwang** [comment]
retest this please.

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/analysis/ResolveOutputRelation.scala (hunk 2 above)
Comment (rdblue) [reply sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/analysis/ResolveOutputRelation.scala]: Were there any modifications other than moving this?

I think that the right way to expose those functions is to move them to a utility class, not to expose this rule itself.

File: sql/core/src/main/scala/org/apache/spark/sql/execution/datasources/rules.scala (hunk 5 above)
Comment (rdblue) [reply sql/core/src/main/scala/org/apache/spark/sql/execution/datasources/rules.scala]: If you aren't going to use the `ResolveOutputRelation` rule, can you explain exactly what the behavior this code path should have and how it differs from `ResolveOutputRelation`?

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/Cast.scala [hunk 10]
```diff
@@ -126,9 +126,12 @@ object Cast {
    */
   def canUpCast(from: DataType, to: DataType): Boolean = (from, to) match {
     case _ if from == to => true
+    case (NullType, _) => false
```
Comment (viirya) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/Cast.scala]: Is this covered by default case previously? Or is it missing before?

File: sql/core/src/main/scala/org/apache/spark/sql/execution/datasources/rules.scala [hunk 11]
```diff
@@ -356,8 +358,28 @@ case class PreprocessTableInsertion(conf: SQLConf) extends Rule[LogicalPlan] {
           s"including ${staticPartCols.size} partition column(s) having constant value(s).")
     }
 
-    val newQuery = DDLPreprocessingUtils.castAndRenameQueryOutput(
```
Comment (viirya) [inline sql/core/src/main/scala/org/apache/spark/sql/execution/datasources/rules.scala]: I saw there is another usage of `castAndRenameQueryOutput` , for `CreateTable` case, should it get rid of unsafe casts too?

File: sql/hive/src/main/scala/org/apache/spark/sql/hive/HiveSessionStateBuilder.scala [hunk 12]
```diff
@@ -81,6 +81,7 @@ class HiveSessionStateBuilder(session: SparkSession, parentState: Option[Session
         RelationConversions(conf, catalog) +:
         PreprocessTableCreation(session) +:
         PreprocessTableInsertion(conf) +:
+        ResolveUpCast +:
```
Comment (viirya) [inline sql/hive/src/main/scala/org/apache/spark/sql/hive/HiveSessionStateBuilder.scala]: Is it good to add a comment like `BaseSessionStateBuilder`?

File: sql/hive/compatibility/src/test/scala/org/apache/spark/sql/hive/execution/HiveCompatibilitySuite.scala (hunk 6 above)
Comment (cloud-fan) [reply sql/hive/compatibility/src/test/scala/org/apache/spark/sql/hive/execution/HiveCompatibilitySuite.scala]: This test is to make sure Spark has the same behavior with Hive, however Hive always add unsafe cast, so we need to turn on this config to make tests pass.

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/analysis/ResolveOutputRelation.scala (hunk 2 above)
Comment (cloud-fan) [reply sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/analysis/ResolveOutputRelation.scala]: I'm OK with moving them to a utility class, but it's better to put analyzer/optimizer rules in its own file, instead of in the `Analyzer` object (can be done in another PR if we decide to create the util class in this PR)

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/Cast.scala (hunk 10 above)
Comment (gengliangwang) [reply sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/Cast.scala]: This is because we need to match it before
```
case (_, StringType) => true
```

File: sql/core/src/main/scala/org/apache/spark/sql/execution/datasources/rules.scala (hunk 11 above)
Comment (gengliangwang) [reply sql/core/src/main/scala/org/apache/spark/sql/execution/datasources/rules.scala]: In `CreateTable`, it should be CTAS and the schema should be the same as the input query.

File: sql/core/src/main/scala/org/apache/spark/sql/execution/datasources/rules.scala (hunk 5 above)
Comment (gengliangwang) [reply sql/core/src/main/scala/org/apache/spark/sql/execution/datasources/rules.scala]: I check and I think it is workable. I will try moving it into the `ResolveOutputRelation` rule.

File: sql/catalyst/src/main/scala/org/apache/spark/sql/types/DecimalType.scala [hunk 13]
```diff
@@ -89,6 +90,7 @@ case class DecimalType(precision: Int, scale: Int) extends FractionalType {
       (precision - scale) <= (dt.precision - dt.scale) && scale <= dt.scale
     case dt: IntegralType =>
       isTighterThan(DecimalType.forType(dt))
+    // For DoubleType/FloatType, the value can be NaN, PositiveInfinity or NegativeInfinity.
```
Comment (cloud-fan) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/types/DecimalType.scala]: For `isWiderThan`, yes it's not safe to cast float/double to decimal because of NaN stuff.

For `isTighterThan`, I think it's safe to cast decimal to float/double if the precision doesn't exceed?

File: sql/catalyst/src/test/scala/org/apache/spark/sql/catalyst/analysis/DataSourceV2AnalysisSuite.scala [hunk 14]
```diff
@@ -290,15 +290,14 @@ abstract class DataSourceV2AnalysisSuite extends AnalysisTest {
       StructField("y", DoubleType))).toAttributes)
 
     val query = TestRelation(StructType(Seq(
-      StructField("x", DoubleType),
+      StructField("x", FloatType),
```
Comment (cloud-fan) [inline sql/catalyst/src/test/scala/org/apache/spark/sql/catalyst/analysis/DataSourceV2AnalysisSuite.scala]: why this change?

File: sql/catalyst/src/test/scala/org/apache/spark/sql/catalyst/analysis/DataSourceV2AnalysisSuite.scala [hunk 15]
```diff
@@ -440,16 +439,15 @@ abstract class DataSourceV2AnalysisSuite extends AnalysisTest {
...
```
Comment (cloud-fan) [inline sql/catalyst/src/test/scala/org/apache/spark/sql/catalyst/analysis/DataSourceV2AnalysisSuite.scala]: ditto

File: sql/catalyst/src/test/scala/org/apache/spark/sql/catalyst/encoders/EncoderResolutionSuite.scala [hunk 16]
```diff
@@ -230,11 +230,12 @@ class EncoderResolutionSuite extends PlanTest {
   castSuccess[Long, String]
   castSuccess[Int, java.math.BigDecimal]
   castSuccess[Long, java.math.BigDecimal]
+  castSuccess[Double, java.math.BigDecimal]
```
Comment (cloud-fan) [inline sql/catalyst/src/test/scala/org/apache/spark/sql/catalyst/encoders/EncoderResolutionSuite.scala]: How is this supported?

File: sql/core/src/main/scala/org/apache/spark/sql/execution/datasources/rules.scala (hunk 5 above)
Comment (cloud-fan) [reply sql/core/src/main/scala/org/apache/spark/sql/execution/datasources/rules.scala]: Have we moved it?

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/Cast.scala [hunk 17]
```diff
@@ -128,8 +128,10 @@ object Cast {
     case _ if from == to => true
     case (from: NumericType, to: DecimalType) if to.isWiderThan(from) => true
     case (from: DecimalType, to: NumericType) if from.isTighterThan(to) => true
-    case (f, t) if legalNumericPrecedence(f, t) => true
+    case (f: NumericType, t: NumericType) if legalNumericPrecedence(f, t) => true
+
     case (DateType, TimestampType) => true
+    case (NullType, _) => false
```
Comment (cloud-fan) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/Cast.scala]: Why can't we upcast null to other nullable types? I think it's pretty to write `INSERT INTO tbl VALUES (1, null)`

File: sql/catalyst/src/main/scala/org/apache/spark/sql/types/DecimalType.scala (hunk 13 above)
Comment (gengliangwang) [reply sql/catalyst/src/main/scala/org/apache/spark/sql/types/DecimalType.scala]: > For isTighterThan, I think it's safe to cast decimal to float/double if the precision doesn't exceed?

Yes, it is. I was about to push the commit to fix tests.

File: sql/catalyst/src/test/scala/org/apache/spark/sql/catalyst/analysis/DataSourceV2AnalysisSuite.scala (hunk 14 above)
Comment (gengliangwang) [reply sql/catalyst/src/test/scala/org/apache/spark/sql/catalyst/analysis/DataSourceV2AnalysisSuite.scala]: In this test case, the error message `"Cannot safely cast", "'x'", "DoubleType to FloatType"` is gone after code changes.
So I think we should make the test case simpler. The nullability error in the `x` column is enough.

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/Cast.scala (hunk 17 above)
Comment (gengliangwang) [reply sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/Cast.scala]: Here we can't know the nullability of the `to` type. We should consider it is not nullable.
For the case you mentioned, it is handled in https://github.com/apache/spark/pull/24806/files#diff-86e655772e8f7cab055d2c2451b52275R134.

File: sql/core/src/main/scala/org/apache/spark/sql/execution/datasources/rules.scala (hunk 5 above)
Comment (gengliangwang) [reply sql/core/src/main/scala/org/apache/spark/sql/execution/datasources/rules.scala]: We need to figure out if we can up-cast decimal to double/float. 
If we can't, then maybe we can't continue this PR.
I have created a new PR https://github.com/apache/spark/pull/24849

**Comment by gengliangwang** [comment]
As discussed in https://github.com/apache/spark/pull/24849 , we can't upcast decimal to double/float. I think we should close this one. What do you think? @cloud-fan @rdblue

File: sql/core/src/main/scala/org/apache/spark/sql/execution/datasources/rules.scala (hunk 5 above)
Comment (rdblue) [reply sql/core/src/main/scala/org/apache/spark/sql/execution/datasources/rules.scala]: @gengliangwang, why is decimal to double/float required? That should not be allowed in either v1 or v2, so I see no reason why we can't use the same rule for both.

File: sql/hive/compatibility/src/test/scala/org/apache/spark/sql/hive/execution/HiveCompatibilitySuite.scala (hunk 6 above)
Comment (rdblue) [reply sql/hive/compatibility/src/test/scala/org/apache/spark/sql/hive/execution/HiveCompatibilitySuite.scala]: Hive automatically inserts unsafe casts?

File: sql/hive/src/test/scala/org/apache/spark/sql/hive/execution/HiveQuerySuite.scala (hunk 7 above)
Comment (rdblue) [reply sql/hive/src/test/scala/org/apache/spark/sql/hive/execution/HiveQuerySuite.scala]: Why not update the test case with the Cast operator?

File: sql/catalyst/src/test/scala/org/apache/spark/sql/catalyst/analysis/DataSourceV2AnalysisSuite.scala (hunk 14 above)
Comment (rdblue) [reply sql/catalyst/src/test/scala/org/apache/spark/sql/catalyst/analysis/DataSourceV2AnalysisSuite.scala]: No, all errors should be reported. Attempting to write a double to a float column is an error that should be shown.

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/Cast.scala (hunk 17 above)
Comment (rdblue) [reply sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/Cast.scala]: I agree with @cloud-fan that this should be allowed. Nullability is an additional check, but the types are compatible.

File: sql/core/src/main/scala/org/apache/spark/sql/execution/datasources/rules.scala (hunk 5 above)
Comment (gengliangwang) [reply sql/core/src/main/scala/org/apache/spark/sql/execution/datasources/rules.scala]: For sql
```
create table t (d double);
insert into t values (10.0);
```
the 10.0 is decimal in Spark SQL parser.
